walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,137 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Arrays;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.Comparator;
 
 public final class FindMeetingQuery {
+  public static final int START_OF_DAY_MINUTES = 0;
+  public static final int END_OF_DAY_MINUTES = 24 * 60; // for a 24 hour day
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    // Sort Collection by start time
+    events.stream().sorted((e1, e2) -> e1.getWhen().start() - e2.getWhen().start());
+
+    Collection<String> optionalAttendees = request.getOptionalAttendees();
+    Collection<String> mandatoryAttendees = request.getAttendees();
+
+    // No meeting attendees: whole day is open
+    if (optionalAttendees.isEmpty() && mandatoryAttendees.isEmpty() && request.getDuration() <= END_OF_DAY_MINUTES) {
+      return Arrays.asList(TimeRange.fromStartEnd(START_OF_DAY_MINUTES, END_OF_DAY_MINUTES, false));
+    }
+
+    // First track times events for all attendees and only mandatory attendees
+    List<TimeRange> allEvents = new ArrayList<>();
+    List<TimeRange> allRequiredEvents = new ArrayList<>();
+    for (Event e : events) {
+      // If the attendee lists are not disjoint (meaning there is at least one attendee in common), 
+      // save this event time
+      if (!Collections.disjoint(e.getAttendees(), mandatoryAttendees)) {
+        // Required for everyone
+        allRequiredEvents.add(e.getWhen());
+        allEvents.add(e.getWhen());
+      } 
+      else if (!Collections.disjoint(e.getAttendees(), optionalAttendees)) {
```

Comment:
No new line between '}' and 'else if'.

https://google.github.io/styleguide/javaguide.html#s4.1.2-blocks-k-r-style